### ISSUE TITLE
distmv.NewNormal does not accept positive semidefinite matrix as Σ

### ISSUE THREAD

**Issue by BiochemLouis** [gonum/gonum#1717]
### What are you trying to do?

I need to draw samples from a multivariate gaussian with a specific μ and Σ, Σ is positive semi-definite, which should be suitable for sampling, but is not allowed in gonum.

### What did you do?

The original draft I got is written in python, up to the normal distribution/sampling, everything works in both numpy and gonum.

Then, using gonum's distmv.NewNormal does not work, while numpy.random.multivariate_normal() can be used to draw the samples.

I can confirm Σ is the same in both.

Looking under the hood, numpy uses SVD to determine if Σ is PSD ([src](https://github.com/numpy/numpy/blob/94010e3e820998afe1cab4d5d915f34552a4476e/numpy/random/mtrand.pyx#L4120)):
```py
(u, s, v) = numpy.linalg.svd(Σ)
psd = np.allclose(np.dot(v.T * s, v), Σ)    
```
Gonum on the other hand uses Cholesky to determine if Σ is PD ([src](https://github.com/gonum/gonum/blob/6b90107d4a29dcf67ae7c214a14935d0bcfbf129/stat/distmv/normal.go#L57)), but the same SVD check works:
```go
svd := new(mat.SVD)
svd.Factorize(Σ, mat.SVDFull)

V := new(mat.Dense)
svd.VTo(V)
s := make([]float64, Σm)
svd.Values(s)

ss := make([]float64, Σm*Σm)
for i := 0; i < n; i++ {
	for j, v := range s {
		ss[i*Σm+j] = v
	}
}
S := mat.NewDense(Σm, Σm, ss)

M := new(mat.Dense)
M.MulElem(V, S)
M.Mul(M, V.T())
psd := mat.EqualApprox(Σ, M, 1e-5)
```

Sorry I cannot share the data or more of the code.

### What did you expect to happen?

distmv.NewNormal should accept positive semidefinite Σ, as far as I understand.

### What actually happened?

distmv.NewNormal accepts only positive definite Σ.

### What version of Go and Gonum are you using?

```
$ go version
go version go1.16.6 linux/amd64
```
```
$ grep "v1/gonum" go.sum 
gonum.org/v1/gonum v0.0.0-20180816165407-929014505bf4/go.mod h1:Y+Yx5eoAFn32cQvJDxZx5Dpnq+c3wtXuadVZAcxbbBo=
gonum.org/v1/gonum v0.8.1-0.20200930085651-eea0b5cb5cc9/go.mod h1:oe/vMfY3deqTw+1EZJhuvEW2iwGF1bW9wwu7XCu0+v0=
gonum.org/v1/gonum v0.8.2/go.mod h1:oe/vMfY3deqTw+1EZJhuvEW2iwGF1bW9wwu7XCu0+v0=
gonum.org/v1/gonum v0.9.3 h1:DnoIG+QAMaF5NvxnGe/oKsgKcAc6PcUyl8q0VetfQ8s=
gonum.org/v1/gonum v0.9.3/go.mod h1:TZumC3NeyVQskjXqmyWt4S3bINhy7B4eYwW69EbyX+0=
```

### Does this issue reproduce with the current master?

It must since the code has not changed.

**Comment by kortschak** [comment]
It's worth noting that [Matlab](https://au.mathworks.com/help/stats/multivariate-normal-distribution.html) disagrees with NumPy here. It is probably worth looking into the consequences of allowing PSD cov in your simulation.

See also [here](https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Drawing_values_from_the_distribution).

**Comment by BiochemLouis** [comment]
Matlab also agrees with numpy on the [mvnrnd page](https://www.mathworks.com/help/stats/mvnrnd.html): 

> Covariances of multivariate normal distributions, specified as a d-by-d symmetric, positive semi-definite matrix or a d-by-d-by-m numeric array.

For R, the mvtnorm package also supports psd by allowing several “methods to be used” ([doc](https://cran.r-project.org/web/packages/mvtnorm/mvtnorm.pdf)):
>method string specifying the matrix decomposition used to determine the matrix root of sigma. Possible methods are eigenvalue decomposition ("eigen", default), singular value decomposition ("svd"), and Cholesky decomposition ("chol"). The Cholesky is typically fastest, not by much though.

As expected, Cholesky does not allow psd, while the other ones do according to [this discussion](https://discourse.julialang.org/t/multivariate-normal-with-positive-semi-definite-covariance-matrix/3029). 

I will try the “hack” from the Julia discussion for now `Σ + 1e-6 * I`, but it seems like psd should work. 

I don’t know how much more computationally expensive eigendecomposition is, but I thought SVD was fairly cheap.

**Comment by kortschak** [comment]
Yes; this is something that we need to discuss. It is unlikely to get into the next release though irrespective of the decision.

**Comment by romanwerpachowski** [comment]
Mathematically, there's nothing wrong with sampling from a multivariate
Gaussian with a PSD covariance matrix, so we should support this. If doing
this has a higher computational cost, then maybe it should be an option one
could choose.

**Comment by kortschak** [comment]
OK. The Julia discussion board were exploring performing the Cholesky decomposition first and then falling back to ED. I think we should avoid ED and use SVD if we go that route, but LDL decomposition seems like it would be a better choice; this would require that Dsptrf be implemented in lapack/gonum (this is not huge at ~500 lines of Fortran in two functions that we don't already have ported — dsptrf and dspr — so it could happen at some time, though time is very unavailable across the project now).

**Comment by vladimir-ch** [comment]
`distmv.Normal` does more than drawing random samples and some of the operations require the covariance matrix to be non-singular. Therefore I think `distmv.Normal` should (or will have to) keep requiring positive definite matrices.

A possibility to address the current request would be to generalize `distmv.NormalRand`
```
func NormalRand(x, mean []float64, chol *mat.Cholesky, src rand.Source) []float64
```
to take an interface instead of `mat.Cholesky`.

> Dsptrf be implemented in lapack/gonum

This should be `Dsytrf`

**Comment by kortschak** [comment]
I think the interface generalisation approach should work well, though perhaps as a new function to avoid the inevitable incorrect use that will come from people not providing correctly conditioned inputs.

> This should be `Dsytrf`

Indeed. Thanks.

**Comment by vladimir-ch** [comment]
> Maybe something like NormalRandCov ...

SGTM

> This should be Dsytrf

I correct myself a bit, Dsytrf would work (it's more general) but for symmetric positive semi-definite matrices Dpstrf should be enough.
